### PR TITLE
add a new option for dynamic profiling of all sites that souper might

### PR DESCRIPTION
**Pull request by regehr**
optimize, not just the ones that it has been able to optimize. this is
useful because it permits us to find the highest profile sites and
then run expensive synthesis only on those, instead of doing synthesis
first.

this profile-all mode needs to run last: it can't run as a peephole
pass.  unfortunately it isn't possible to make a single pass that is
scheduled either as a peephole or running last, so this patch compiles
two versions of the pass and then sclang chooses the right one

also, it turns out that running profile code from an LLVM function
pass was incorrect and only happened to work: it is illegal to add new
functions from a function pass. so turn the function pass into a
module pass.

finally, fix a different bug in the profile instrumentation where we
had failed to put the atomic increment below any phi nodes that happen
to be there.

### CODE REVIEW

File: lib/Pass/Pass.cpp [hunk 1]
```diff
@@ -133,27 +150,32 @@ struct SouperPass : public FunctionPass {
...
     InstContext IC;
     ExprBuilderContext EBC;
     CandidateMap CandMap;
-
-    LoopInfo *LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
-
-    FunctionCandidateSet CS = ExtractCandidatesFromPass(&F, LI, IC, EBC);
+    LoopInfo *LI = &getAnalysis<LoopInfoWrapperPass>(*F).getLoopInfo();
+    if (!LI)
+      report_fatal_error("foo!");
```
Comment (rsas) [inline lib/Pass/Pass.cpp]: Is there a better error message?

File: lib/Pass/Pass.cpp [hunk 2]
```diff
@@ -133,27 +150,32 @@ struct SouperPass : public FunctionPass {
...
     new AtomicRMWInst(AtomicRMWInst::Add, CntVar,
                       ConstantInt::get(C, APInt(64, 1)), Monotonic, CrossThread,
                       BI);
   }
 
-  bool runOnFunction(Function &F) {
+  std::set <Function *> DoneFuncs;
+
+  bool runOnFunction(Function *F) {
     bool changed = false;
```
Comment (rsas) [inline lib/Pass/Pass.cpp]: Can't remember if there was a naming policy for local vars, shouldn't it be ``Changed``?

File: lib/Pass/Pass.cpp [hunk 3]
```diff
@@ -250,6 +269,20 @@ struct SouperPass : public FunctionPass {
 
     return changed;
   }
+
+  bool runOnModule(Module &M) {
+    bool changed = false;
```
Comment (rsas) [inline lib/Pass/Pass.cpp]: Ditto.

**Comment by rsas** [comment]
LGTM, but couldn't test it yet.

File: lib/Pass/Pass.cpp (hunk 1 above)
Comment (regehr) [inline lib/Pass/Pass.cpp]: Ouch how did that slip through??

**Comment by regehr** [comment]
Issues fixed-- ok to merge?

File: lib/Pass/Pass.cpp [hunk 4]
```diff
@@ -133,27 +150,32 @@ struct SouperPass : public FunctionPass {
...
     std::string FunctionName;
-    if (F.hasLocalLinkage()) {
+    if (F->hasLocalLinkage()) {
       FunctionName =
-        (F.getParent()->getModuleIdentifier() + ":" + F.getName()).str();
+        (F->getParent()->getModuleIdentifier() + ":" + F->getName()).str();
     } else {
-      FunctionName = F.getName();
+      FunctionName = F->getName();
     }
```
Comment (rsas) [inline lib/Pass/Pass.cpp]: Are we still following the code style policy "no curly braces for single lines"? If yes, please edit both ``if`` and ``else`` path.

**Comment by rsas** [comment]
Except of the code style question, LGTM.

File: lib/Pass/Pass.cpp (hunk 4 above)
Comment (regehr) [inline lib/Pass/Pass.cpp]: Overall I don't think we're very consistent about this choice. We can address this later with a cleanup patch (though I don't really have strong feelings about doing this) but I think for now this is fine. Thanks for the careful read!

**Pull request merged**